runtime/vm/resolvesupport.cpp

```diff
@@ -1588,6 +1588,36 @@ resolveMethodTypeRefInto(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIn
 		}
 	}
 
+	/* perform visibility checks for the returnType and all parameters */
+	if (methodType != NULL) {
+		/* check returnType */
+		J9Class *senderClass = ramCP->ramClass;
+		J9Class *returnTypeClass = J9VM_J9CLASS_FROM_HEAPCLASS(vmThread, J9VMJAVALANGINVOKEMETHODTYPE_RETURNTYPE(vmThread, methodType));
+		bool illegalAccessError = false;
+
+		if (J9_VISIBILITY_ALLOWED != checkVisibility(vmThread, senderClass, returnTypeClass, returnTypeClass->romClass->modifiers, resolveFlags)) {
+			illegalAccessError = true;
+		} else {
+			/* check paramTypes */
+			j9object_t argTypesObject = J9VMJAVALANGINVOKEMETHODTYPE_ARGUMENTS(vmThread, methodType);
+			U_32 typeCount = J9INDEXABLEOBJECT_SIZE(vmThread, argTypesObject);
+			for (UDATA i = 0; i < typeCount; i++) {
+				J9Class *paramClass = J9VM_J9CLASS_FROM_HEAPCLASS(vmThread, J9JAVAARRAYOFOBJECT_LOAD(vmThread, argTypesObject, i));
+
+				if (J9_VISIBILITY_ALLOWED != checkVisibility(vmThread, senderClass, paramClass, paramClass->romClass->modifiers, resolveFlags)) {
+					illegalAccessError = true;
+				}
+			}
+		}
+		if (illegalAccessError) {
+			j9object_t lookupSigString = vmThread->javaVM->memoryManagerFunctions->j9gc_createJavaLangString(vmThread, J9UTF8_DATA(lookupSig), J9UTF8_LENGTH(lookupSig), 0);
```

Comment:
Can a more specific error message be added?  Ideally highlighting the class that fails the access check?

And there should be a tracepoint here as well.

Reply:
yes, I put a note in the commit saying I will add better error messages in the next PR.